### ISSUE TITLE
No metrics are being collected

### ISSUE THREAD

**Issue by DariusSlowik** [nawafalageel/docker_container_gpu_exporter#1]
I am running the script on Debian 12 x86_64.
The http server starts at http://localhost:9066.
However, when I check the browser or curl the address no data are visible.
![image](https://github.com/user-attachments/assets/b63c6f61-6123-43dd-8c79-9d6c04d3d628)

Process starts as well, but one of subprocesses (red one) ends immediately
![image](https://github.com/user-attachments/assets/39b9f0c8-baae-4926-9a69-c01eed4c945e)

Prometheus sees the job correctly as well
![image](https://github.com/user-attachments/assets/273c598b-c0d5-4701-9cb2-54d7e0ff489c)

My neofetch output:
![image](https://github.com/user-attachments/assets/8e0761d6-52ef-4773-9181-bc2360ea6217)

nvidia-smi output:
![image](https://github.com/user-attachments/assets/d7f8ea41-9dca-4103-b0e4-f01c6f811231)

If you still up to it, could I get any help?

**Comment by nawafalageel** [comment]
Hiii, 

It's been a while since I last encountered an issue with the deployed scripts. However, I noticed that you're using Debian 12 x86_64, whereas I'm using Ubuntu 18.04. This could be a key difference I can see. I'd be happy to help you debug the problem. 👍

**Comment by DariusSlowik** [comment]
Hello,
honestly I didn't even think you'd answer :P
Thank you :)

At first I thought, maybe containerisation would fix the problem?
so I succesfully created an image and container (it turns out, version of python image needs to be 3.11< [`python:3.10`], also I applied newest versions of `PyYAML` and `prometheus-client` in `requirements.txt`), but the results were the same. Metrics were not displayed.

So I tried again with the service within my own system. I checked if I have prometheus-client and I did but it turned out it was wrongly installed and I was suggested to create python virtual environment with venv.
NOTE! Debian prevents mixing packages installed with apt and pip, that's why virtual env is needed.

```
python3 -m venv .venv
source .venv/bin/activate
```

With those above I entered virtual environment and installed requirements again

```
pip3 install prometheus-client
pip3 install PyYAML
```

Then I ran the code again with `python3 docker_gpu_exporter_v3.py`
And metrics appeared.
![image](https://github.com/user-attachments/assets/78851dff-704d-45eb-a064-7c32f359e0a1)

I tried to apply those fixes within container image and added some lines in Dockerfile but it still didn't work. However my skills with docker are not perfect therefore I might've applied it incorrectly.
![image](https://github.com/user-attachments/assets/a33ed127-cd6f-466f-a4d0-25ba723257dd)

Here a screenshot of the Current Dockerfile after virtualisation:
![image](https://github.com/user-attachments/assets/c32f906f-2279-44da-b0f0-5dcdb53a9462)

Maybe linking ubuntu image instead of python image would fix the problem here?

**Comment by DariusSlowik** [comment]
I spent some time on this and I think I figured it out.
This was my fault (but not all). Linking with system processes and components was incorrect.
The `Dockerfile` in the main repo is fine as is, no need to change or create virtual env in the image, also latest ver of python started working as well.
During `docker run` command, proper additional flags are needed:
`--privileged` - to link host devices.
It links all devices but linking only nvidia ones would be enough I think: `-v /dev/<nvidia-device>:/dev/<nvidia-device>`
Without it container doesn't see NVML libraries.

`--pid` - to link host processes. Without it nvidia-smi returns empty table of processes.

`--gpus` - to link graphics drivers. Without it scripts within container can't use nvidia-smi.

`-v /usr/bin/docker:/usr/bin/docker` - additional linking of docker commands. Without it and the other one mentioned in `README.md`, `docker ps` in shell script within container doesn't work.

```
docker build -t docker-container-gpu-exporter .
```
```
docker run --name="docker-container-gpu-exporter"
                  --privileged \
                  --gpus all \
                  --pid=host \
                  --restart=unless-stopped
                  -v /var/run/docker.sock:/var/run/docker.sock \
                  -v /usr/bin/docker:/usr/bin/docker \
                  -p 9066:9066 \
                  -d docker-container-gpu-exporter
```
Now metrics are being collected by container :)

![image](https://github.com/user-attachments/assets/c5f30b48-c4de-46fb-9901-2e2801d58c56)

![image](https://github.com/user-attachments/assets/dcb51b0f-2764-47c0-8604-5154686127aa)
